### ISSUE TITLE
Server crashes right after putting in name

### ISSUE THREAD

**Issue by davidcernat** [TES3MP/TES3MP#80]
Recently, the server has started crashing right after players put in their name. It only happens sometimes, but is clearly the most major crash that now happens in tes3mp.

The problem doesn't seem related to Lua scripts, because there is no Lua error of any kind in the server's console.

I would have expected changes in the past week to be behind the crash, but Luc says he has experienced this with the build from the end of last month as well.

Observations so far:

1) It doesn't matter how many players are on the server. I've gotten the crash when joining a newly started server by myself.

2) It doesn't matter how long the server has been on.

**Comment by davidcernat** [comment]
Here's the conversation I had with Koncord about this:

> 8:28 PM - David C.: Sometimes, a player joins the server, and the server gets this far for the new player in Networking:
> 8:28 PM - David C.: https://github.com/TES3MP/openmw-tes3mp/blob/master/apps/openmw-mp/Networking.cpp#L122
> 8:28 PM - David C.: So the server starts sending the new player information about the other players:
> 8:28 PM - David C.: https://github.com/TES3MP/openmw-tes3mp/blob/master/apps/openmw-mp/Networking.cpp#L518
> 8:30 PM - David C.: However, it iterates over an invalid player...
> 8:30 PM - David C.: And crashes.
> 8:30 PM - David C.: Here is some debug I added:
> 8:30 PM - David C.: https://gist.github.com/davidcernat/fbba3ea6f452f9a4ea725f0d1565c9cf
> 8:31 PM - David C.: Here is an example of this debug when a player manages to connect and there is no problem:
> 8:31 PM - David C.: [2016-10-24 19:34:29] [INFO]: Start of NewPlayer with guid 2454313235
> [2016-10-24 19:34:29] [WARN]: Preparing to send other players info to new guy
> [2016-10-24 19:34:29] [WARN]: Checking 578963370 for new guy
> [2016-10-24 19:34:29] [WARN]: Sending info of 578963370 to the new guy
> [2016-10-24 19:34:29] [WARN]: Checking 3644116260 for new guy
> [2016-10-24 19:34:29] [WARN]: Sending info of 3644116260 to the new guy
> [2016-10-24 19:34:29] [WARN]: Checking 2454313235 for new guy
> [2016-10-24 19:34:29] [WARN]: Victory!
> 8:31 PM - David C.: Here is an example of this debug when there is a crash:
> 8:32 PM - David C.: [2016-10-24 19:38:53] [INFO]: Start of NewPlayer with guid 60276922
> [2016-10-24 19:38:53] [WARN]: Preparing to send other players info to new guy
> [2016-10-24 19:38:53] [WARN]: Checking 578963370 for new guy
> [2016-10-24 19:38:53] [WARN]: Sending info of 578963370 to the new guy
> [2016-10-24 19:38:53] [WARN]: Checking 60276922 for new guy
> [2016-10-24 19:38:53] [WARN]: Checking 2454313235 for new guy
> [2016-10-24 19:38:53] [WARN]: Sending info of 2454313235 to the new guy
> [2016-10-24 19:38:53] [WARN]: Checking 4294967295 for new guy
> [2016-10-24 19:38:53] [WARN]: Sending info of 4294967295 to the new guy
> 8:32 PM - David C.: That's exactly when it crashed.
> 8:32 PM - David C.: Here is another crash:
> 8:32 PM - David C.: [2016-10-24 19:11:03] [INFO]: Start of NewPlayer with guid 351107079
> [2016-10-24 19:11:03] [WARN]: Preparing to send other players info to new guy
> [2016-10-24 19:11:03] [WARN]: Checking 351107079 for new guy
> [2016-10-24 19:11:03] [WARN]: Checking 284336979 for new guy
> [2016-10-24 19:11:03] [WARN]: Sending info of 284336979 to the new guy
> [2016-10-24 19:11:03] [WARN]: Checking 4294967295 for new guy
> [2016-10-24 19:11:03] [WARN]: Sending info of 4294967295 to the new guy
> 8:33 PM - David C.: According to Wikipedia:
> 8:33 PM - Koncord: how to reproduce bug? I'll check it tomorrow.
> 8:33 PM - David C.: "The number 4,294,967,295, equivalent to the hexadecimal value FFFF,FFFF16, is the maximum value for a 32-bit unsigned integer in computing."
> 8:33 PM - David C.: It could be memory corruption only on Windows.
> 8:34 PM - David C.: On Windows, you just log in enough times and it happens.
> 8:34 PM - David C.: First crash happened after 5 log ins.
> 8:34 PM - David C.: Second crash happened after 9 log ins.
> 8:34 PM - David C.: We had many many other crashes, but the debug for them was not good enough.
> 8:35 PM - David C.: Maybe the way players are deleted on the server is not memory safe:
> 8:36 PM - David C.: https://github.com/TES3MP/openmw-tes3mp/blob/master/apps/openmw-mp/Player.cpp#L22
> 8:37 PM - David C.: Maybe this condition just needs to be better so it ignores FFFFFF:
> 8:37 PM - David C.: for (TPlayers::iterator pl = players->begin(); pl != players->end(); pl++)
> 8:39 PM - Koncord: I can debug this bug on VirtualMachine tomorrow

**Comment by davidcernat** [comment]
Some more info I relayed to Koncord in a conversation:

> 11:50 AM - David C.: I added some extra debug to tes3mp-server's Player.cpp
> 11:50 AM - David C.: When a player is deleted, the remaining size of the players map is printed in the console.
> 11:51 AM - David C.: The code looks like this now:

```
    if (players[id] != 0)
    {
        LOG_APPEND(Log::LOG_INFO, "- Emptying slot %i",
            players[id]->GetID());

        slots[players[id]->GetID()] = 0;
        delete players[id];
        players.erase(id);
    }

    LOG_APPEND(Log::LOG_INFO, "- %i remaining",
        players.size());
```

> 11:51 AM - David C.: Well, I can now predict when a server crash will happen.
> 11:51 AM - David C.: It happens when everyone has left the server and the server console says:
> 11:51 AM - David C.: - 1 remaining
> 11:52 AM - David C.: If someone joins after that, the server crashes.

**Comment by davidcernat** [comment]
Fixed by 3c88f6f0cd571389dc111e7325c0dd26086a656e.
